### PR TITLE
Modular commands attempt 3 #943

### LINKED ISSUE #943: Allow modular command definition
Allow commands to be split into multiple classes that are registered with cmd2 instead of requiring all commands to be defined in one giant class.

Breakdown:
- Define a base class for these satellite command definitions.
- Everywhere in cmd2 where it introspects for `do_`, `complete_`, or `help_` also search in the satellite command objects.
- Stretch - auto-detect all satellite command definitions (by subclass) and load them automatically on init of cmd2.Cmd.
- Example of loading a satellite command object

### PR DESCRIPTION
**Pull request by anselor**
ok, this should be the real one we want.

Going back to being integrated into the cmd2 core instead of a plugin.
Removed support for arbitrary functions as loadable modular commands. Modular commands must be in a CommandSet.
Added additional documentation and examples.

Closes #943

### CODE REVIEW

File: cmd2/cmd2.py [hunk 1]
```diff
@@ -381,6 +399,218 @@ def __init__(self, completekey: str = 'tab', stdin=None, stdout=None, *,
...
+            command_func = self.cmd_func(command_name)
+            if command_func is None or not hasattr(command_func, constants.CMD_ATTR_ARGPARSER):
+                raise TypeError('Could not find command "{}" needed by sub-command: {}'
+                                .format(command_name, str(method)))
+            command_parser = getattr(command_func, constants.CMD_ATTR_ARGPARSER)
+            if command_parser is None:
+                raise TypeError('Could not find argparser for command "{}" needed by sub-command: {}'
+                                .format(command_name, str(method)))
+
+            if hasattr(method, '__doc__') and method.__doc__ is not None:
```
Comment (kmvanbrunt) [inline cmd2/cmd2.py:562]: The `as_subcommand_to` decorator should take a `help` argument instead of using the methods docstring.

File: cmd2/cmd2.py [hunk 2]
```diff
@@ -130,7 +134,9 @@ def __init__(self, completekey: str = 'tab', stdin=None, stdout=None, *,
                  startup_script: str = '', use_ipython: bool = False,
                  allow_cli_args: bool = True, transcript_files: Optional[List[str]] = None,
                  allow_redirection: bool = True, multiline_commands: Optional[List[str]] = None,
-                 terminators: Optional[List[str]] = None, shortcuts: Optional[Dict[str, str]] = None) -> None:
+                 terminators: Optional[List[str]] = None, shortcuts: Optional[Dict[str, str]] = None,
+                 command_sets: Optional[Iterable[CommandSet]] = None,
+                 auto_load_commands: bool = True) -> None:
```
[the remark this answers is not in the document]
Comment (anselor) [reply cmd2/cmd2.py:139]: We do because, for projects that choose to use CommandSets they will 'just work'. The search for child classes is fast because it's asking Python for child classes that have already been loaded. This should just be a lookup of information that's already loaded in memory.

File: cmd2/cmd2.py [hunk 3]
```diff
@@ -381,6 +399,218 @@ def __init__(self, completekey: str = 'tab', stdin=None, stdout=None, *,
...
+
+                if hasattr(self, COMPLETER_FUNC_PREFIX + cmd_name):
+                    delattr(self, COMPLETER_FUNC_PREFIX + cmd_name)
+                if hasattr(self, HELP_FUNC_PREFIX + cmd_name):
+                    delattr(self, HELP_FUNC_PREFIX + cmd_name)
+
+            cmdset.on_unregister(self)
+            self._installed_command_sets.remove(cmdset)
+
+    def _register_subcommands(self, cmdset: Union[CommandSet, 'Cmd']) -> None:
```
[the remark this answers is not in the document]
Comment (anselor) [reply cmd2/cmd2.py:528]: This allows defining sub-commands using this decorator directly in a Cmd child class.

File: cmd2/argparse_custom.py [hunk 4]
```diff
@@ -724,21 +724,70 @@ def _format_args(self, action, default_metavar) -> str:
...
 # noinspection PyCompatibility
 class Cmd2ArgumentParser(argparse.ArgumentParser):
     """Custom ArgumentParser class that improves error and help output"""
 
-    def __init__(self, *args, **kwargs) -> None:
-        if 'formatter_class' not in kwargs:
-            kwargs['formatter_class'] = Cmd2HelpFormatter
-
-        super().__init__(*args, **kwargs)
+    def __init__(self,
```
[the remark this answers is not in the document]
Comment (anselor) [reply cmd2/argparse_custom.py:749]: 1) It's really annoying to use because all of the IDE help is lost with *args/**kwargs. Also, this is something that @tleonhardt identified as an antipattern on one of the previous iterations.
2) The way python default values work, a list declared as a default value is shared across all instantiations of an object. If it is subsequently modified it is modified for all instances. Generally considered unsafe.

File: cmd2/argparse_custom.py [hunk 5]
```diff
@@ -724,21 +724,70 @@ def _format_args(self, action, default_metavar) -> str:
...
+            allow_abbrev=allow_abbrev)
+        self.register('action', 'unloadable_parsers', _UnloadableSubParsersAction)
+
+    def add_subparsers(self, unloadable=False, **kwargs):
+        """
+        Custom override. Sets a default title if one was not given.
 
-    def add_subparsers(self, **kwargs):
-        """Custom override. Sets a default title if one was not given."""
+        :param unloadable: Flag whether this sub-parsers group will support unloading parsers
```
[the remark this answers is not in the document]
Comment (anselor) [reply cmd2/argparse_custom.py:781]: Mostly to minimize potential impact if you don't need this feature. I can go either way.

File: cmd2/cmd2.py [hunk 6]
```diff
@@ -572,18 +569,18 @@ def _register_subcommands(self, cmdset: Union[CommandSet, 'Cmd']) -> None:
 
             for action in command_parser._actions:
                 if isinstance(action, _UnloadableSubParsersAction):
-                    action.add_parser(subcommand_name, parents=[subcmd_parser], help=help_text)
+                    action.add_parser(subcommand_name, parents=[subcmd_parser], help=help_text, aliases=aliases)
```
Comment (kmvanbrunt) [inline cmd2/cmd2.py:572]: I don't think you should pass `help` and `aliases` if they are `None`.
Look how [argparse](https://github.com/python/cpython/blob/3.8/Lib/argparse.py#L1109) uses these arguments.

I would suggest doing:

```Python
kwargs = dict()
if help_text is not None:
    kwargs['help'] = help_text
if aliases is not None:
    kwargs['aliases '] = aliases

action.add_parser(subcommand_name, parents=[subcmd_parser], **kwargs)
```

File: cmd2/decorators.py [hunk 7]
```diff
@@ -288,3 +335,43 @@ def cmd_wrapper(cmd2_app, statement: Union[Statement, str], **kwargs: Dict[str,
...
+        if parser.description is None and func.__doc__:
+            parser.description = func.__doc__
+
+        parser.set_defaults(func=func)
+
+        # Set some custom attributes for this command
+        setattr(func, constants.SUBCMD_ATTR_COMMAND, command)
+        setattr(func, constants.CMD_ATTR_ARGPARSER, parser)
+        setattr(func, constants.SUBCMD_ATTR_NAME, subcommand)
+        if help_text:
```
Comment (kmvanbrunt) [inline cmd2/decorators.py:369]: Should be: `if help_text is not None:`

File: cmd2/decorators.py [hunk 8]
```diff
@@ -288,3 +335,43 @@ def cmd_wrapper(cmd2_app, statement: Union[Statement, str], **kwargs: Dict[str,
...
+            setattr(func, constants.SUBCMD_ATTR_HELP, help_text)
+        if aliases:
```
Comment (kmvanbrunt) [inline cmd2/decorators.py:371]: Should be: `if aliases is not None:`

File: cmd2/decorators.py [hunk 9]
```diff
@@ -1,12 +1,15 @@
...
 """Decorators for ``cmd2`` commands"""
 import argparse
-from typing import Any, Callable, Dict, List, Optional, Union
+from typing import TYPE_CHECKING, Any, Callable, Dict, Iterable, List, Optional, Tuple, Union
 
 from . import constants
 from .exceptions import Cmd2ArgparseError
 from .parsing import Statement
 
+if TYPE_CHECKING:
```
Comment (tleonhardt) [inline cmd2/decorators.py:10]: Recommend adding pragma to except this `if` from code coverage analysis

File: cmd2/decorators.py [hunk 10]
```diff
@@ -30,6 +33,42 @@ def cat_decorator(func):
...
+    :arg args: The positional arguments to inspect
+    :return: The cmd2.Cmd reference and the command line statement
+    """
+    for pos, arg in enumerate(args):
+        from cmd2 import Cmd
+        if isinstance(arg, Cmd) and len(args) > pos:
+            next_arg = args[pos + 1]
+            if isinstance(next_arg, (Statement, str)):
+                return arg, args[pos + 1]
+    raise TypeError('Expected arguments: cmd: cmd2.Cmd, statement: Union[Statement, str] Not found')
```
Comment (tleonhardt) [inline cmd2/decorators.py:56]: Recommend adding a unit test for the case where this `raise` is hit

File: cmd2/decorators.py [hunk 11]
```diff
@@ -288,3 +335,45 @@ def cmd_wrapper(cmd2_app, statement: Union[Statement, str], **kwargs: Dict[str,
...
+
+        parser.set_defaults(func=func)
+
+        # Set some custom attributes for this command
+        setattr(func, constants.SUBCMD_ATTR_COMMAND, command)
+        setattr(func, constants.CMD_ATTR_ARGPARSER, parser)
+        setattr(func, constants.SUBCMD_ATTR_NAME, subcommand)
+        parser_args = {}
+        if help_text is not None:
+            parser_args['help'] = help_text
```
Comment (tleonhardt) [inline cmd2/decorators.py:371]: Recommend adding unit test for case `help_text` is `None`

File: cmd2/decorators.py [hunk 12]
```diff
@@ -288,3 +335,45 @@ def cmd_wrapper(cmd2_app, statement: Union[Statement, str], **kwargs: Dict[str,
...
+        if aliases is not None:
+            parser_args['aliases'] = aliases[:]
```
Comment (tleonhardt) [inline cmd2/decorators.py:373]: Recommend adding unit test for case `aliases` is `None`

File: cmd2/argparse_custom.py [hunk 13]
```diff
@@ -528,6 +532,39 @@ def _match_argument_wrapper(self, action, arg_strings_pattern) -> int:
...
+
+    This is a custom method being added to the argparse._SubParsersAction
+    class so cmd2 can remove subcommands from a parser.
+
+    :param self: instance of the _SubParsersAction being edited
+    :param name: name of the sub-parser to remove
+    """
+    for choice_action in self._choices_actions:
+        if choice_action.dest == name:
+            self._choices_actions.remove(choice_action)
```
Comment (tleonhardt) [inline cmd2/argparse_custom.py:552]: Recommend adding unit test where the code inside this `for` and inner `if` is executed

File: cmd2/cmd2.py [hunk 14]
```diff
@@ -381,6 +399,229 @@ def __init__(self, completekey: str = 'tab', stdin=None, stdout=None, *,
...
+        """Load modular command definitions."""
+        # Search for all subclasses of CommandSet, instantiate them if they weren't provided in the constructor
+        all_commandset_defs = CommandSet.__subclasses__()
+        existing_commandset_types = [type(command_set) for command_set in self._installed_command_sets]
+        for cmdset_type in all_commandset_defs:
+            init_sig = inspect.signature(cmdset_type.__init__)
+            if cmdset_type in existing_commandset_types or \
+                    len(init_sig.parameters) != 1 or \
+                    'self' not in init_sig.parameters:
+                continue
```
Comment (tleonhardt) [inline cmd2/cmd2.py:412]: Recommend adding a unit test to exercise the case where this `continue` is reached

File: cmd2/cmd2.py [hunk 15]
```diff
@@ -381,6 +399,229 @@ def __init__(self, completekey: str = 'tab', stdin=None, stdout=None, *,
...
+                predicate=lambda meth: isinstance(meth, Callable)
+                and hasattr(meth, '__name__') and meth.__name__.startswith(COMMAND_FUNC_PREFIX))
+
+            for method in methods:
+                cmd_name = method[0][len(COMMAND_FUNC_PREFIX):]
+
+                # Enable the command before uninstalling it to make sure we remove both
+                # the real functions and the ones used by the DisabledCommand object.
+                if cmd_name in self.disabled_commands:
+                    self.enable_command(cmd_name)
```
Comment (tleonhardt) [inline cmd2/cmd2.py:523]: Recommend adding a unit test for the case where the code within this `if` is it because the command is in `disabled_commands`

File: cmd2/cmd2.py [hunk 16]
```diff
@@ -381,6 +399,229 @@ def __init__(self, completekey: str = 'tab', stdin=None, stdout=None, *,
...
+        # iterate through all matching methods
+        for method_name, method in methods:
+            subcommand_name = getattr(method, constants.SUBCMD_ATTR_NAME)
+            command_name = getattr(method, constants.SUBCMD_ATTR_COMMAND)
+            subcmd_parser = getattr(method, constants.CMD_ATTR_ARGPARSER)
+            parser_args = getattr(method, constants.SUBCMD_ATTR_PARSER_ARGS, {})
+
+            # Search for the base command function and verify it has an argparser defined
+            if command_name in self.disabled_commands:
+                command_func = self.disabled_commands[command_name].command_function
```
Comment (tleonhardt) [inline cmd2/cmd2.py:562]: Please add unit test to exercise this case

File: cmd2/cmd2.py [hunk 17]
```diff
@@ -381,6 +399,229 @@ def __init__(self, completekey: str = 'tab', stdin=None, stdout=None, *,
...
+                                .format(command_name, str(method)))
+            command_parser = getattr(command_func, constants.CMD_ATTR_ARGPARSER, None)
+            if command_parser is None:
+                raise TypeError('Could not find argparser for command "{}" needed by subcommand: {}'
+                                .format(command_name, str(method)))
+
+            if isinstance(cmdset, CommandSet):
+                command_handler = _partial_passthru(method, self)
+            else:
+                command_handler = method
```
Comment (tleonhardt) [inline cmd2/cmd2.py:577]: Please add unit test to exercise this `else` case

File: cmd2/cmd2.py [hunk 18]
```diff
@@ -381,6 +399,229 @@ def __init__(self, completekey: str = 'tab', stdin=None, stdout=None, *,
...
+        )
+
+        # iterate through all matching methods
+        for method_name, method in methods:
+            subcommand_name = getattr(method, constants.SUBCMD_ATTR_NAME)
+            command_name = getattr(method, constants.SUBCMD_ATTR_COMMAND)
+
+            # Search for the base command function and verify it has an argparser defined
+            if command_name in self.disabled_commands:
+                command_func = self.disabled_commands[command_name].command_function
```
Comment (tleonhardt) [inline cmd2/cmd2.py:609]: Please add unit test for this `if`

File: cmd2/cmd2.py [hunk 19]
```diff
@@ -381,6 +399,229 @@ def __init__(self, completekey: str = 'tab', stdin=None, stdout=None, *,
...
+            else:
+                command_func = self.cmd_func(command_name)
+
+            if command_func is None:
+                raise TypeError('Could not find command "{}" needed by subcommand: {}'
```
Comment (tleonhardt) [inline cmd2/cmd2.py:614]: Please add unit test for this if

File: cmd2/cmd2.py [hunk 20]
```diff
@@ -381,6 +399,229 @@ def __init__(self, completekey: str = 'tab', stdin=None, stdout=None, *,
...
+                                .format(command_name, str(method)))
+            command_parser = getattr(command_func, constants.CMD_ATTR_ARGPARSER, None)
+            if command_parser is None:
+                raise TypeError('Could not find argparser for command "{}" needed by subcommand: {}'
```
Comment (tleonhardt) [inline cmd2/cmd2.py:618]: Please add unit test for this if

File: cmd2/argparse_custom.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (tleonhardt) [reply cmd2/argparse_custom.py:768]: I really like having good help in PyCharm or VSCode.

File: cmd2/argparse_completer.py [hunk 21]
```diff
@@ -181,7 +182,8 @@ def __init__(self, parser: argparse.ArgumentParser, cmd2_app: cmd2.Cmd, *,
                 if isinstance(action, argparse._SubParsersAction):
                     self._subcommand_action = action
 
-    def complete_command(self, tokens: List[str], text: str, line: str, begidx: int, endidx: int) -> List[str]:
+    def complete_command(self, tokens: List[str], text: str, line: str, begidx: int, endidx: int, *,
+                         cmd_set: CommandSet = None) -> List[str]:
```
Comment (kmvanbrunt) [inline cmd2/argparse_completer.py:186]: Change type hint to: `cmd_set: Optional[CommandSet] = None`

File: cmd2/argparse_completer.py [hunk 22]
```diff
@@ -543,7 +547,8 @@ def format_help(self, tokens: List[str]) -> str:
 
     def _complete_for_arg(self, arg_action: argparse.Action,
                           text: str, line: str, begidx: int, endidx: int,
-                          consumed_arg_values: Dict[str, List[str]]) -> List[str]:
+                          consumed_arg_values: Dict[str, List[str]], *,
+                          cmd_set: CommandSet = None) -> List[str]:
```
Comment (kmvanbrunt) [inline cmd2/argparse_completer.py:551]: Change type hint to: cmd_set: Optional[CommandSet] = None

File: cmd2/argparse_custom.py (hunk 13 above)
Comment (anselor) [reply cmd2/argparse_custom.py:552]: As far as I can tell this is a bug in how coverage is being reported. I debugged and stepped through this and verified it is in fact getting hit.

File: cmd2/cmd2.py [hunk 23]
```diff
@@ -381,6 +404,288 @@ def __init__(self, completekey: str = 'tab', stdin=None, stdout=None, *,
...
+            command_parser = getattr(command_func, constants.CMD_ATTR_ARGPARSER, None)
+            if command_parser is None:
+                raise CommandSetRegistrationError('Could not find argparser for command "{}" needed by subcommand: {}'
+                                                  .format(command_name, str(method)))
+
+            if isinstance(cmdset, CommandSet):
+                command_handler = _partial_passthru(method, self)
+            else:
+                command_handler = method
+            subcmd_parser.set_defaults(handler=command_handler)
```
Comment (kmvanbrunt) [inline cmd2/cmd2.py:624]: I think `handler` should be renamed to something more unique since it's possible for a user to already have an argument called handler.

Make this name a constant so users can just do: 
`func = getattr(ns, CONSTANT, None)`
